File: Python/735. Asteroid Collision.py
from typing import List


# noinspection PyMethodMayBeStatic
class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        stack = []

        for asteroid in asteroids:
            if asteroid > 0:
                stack.append(asteroid)
            else:
                abs_asteroid = abs(asteroid)

                while len(stack) > 0 and 0 < stack[-1] < abs_asteroid:
                    stack.pop()

                end_stack = stack[-1] if len(stack) > 0 else None
                stack_length = len(stack)

                if stack_length > 0 and end_stack == abs_asteroid:
                    stack.pop()
                elif stack_length == 0 or end_stack < 0:
                    stack.append(asteroid)

        return stack

File: PHP/735. Asteroid Collision.php
<?php

/**
 * @noinspection PhpIllegalPsrClassPathInspection
 * @noinspection PhpMultipleClassDeclarationsInspection
 * @noinspection PhpUnused
 */
class Solution
{
    /**
     * @param  int[]  $asteroids
     * @return int[]
     */
    public function asteroidCollision(array $asteroids): array
    {
        $stack = [];

        for ($i = 0; $i < count($asteroids); $i++) {
            $current = $asteroids[$i];

            if ($current > 0) {
                $stack[] = $current;

                continue;
            }

            $absCurrent = abs($current);

            while (count($stack) > 0 && end($stack) > 0 && end($stack) < $absCurrent) {
                array_pop($stack);
            }

            $endStack = end($stack);
            $stackLength = count($stack);

            if ($stackLength > 0 && $endStack === $absCurrent) {
                array_pop($stack);
            } elseif ($stackLength === 0 || $endStack < 0) {
                $stack[] = $current;
            }
        }

        return $stack;
    }
}
